src/metadata.rs

```diff
@@ -0,0 +1,161 @@
+use std::collections::HashMap;
+
+use crate::{
+    account::{account, Account},
+    amount::{amount, currency, Amount},
+    date::{date, Date},
+    string::string,
+};
+
+use nom::{
+    branch::alt,
+    bytes::complete::tag,
+    character::complete::{line_ending, one_of, satisfy, space0, space1},
+    combinator::{map, recognize},
+    multi::{fold_many0, many0},
+    sequence::{pair, preceded, separated_pair, tuple},
+    IResult,
+};
+
+pub type Metadata<'a> = HashMap<String, Value<'a>>;
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum Value<'a> {
+    Account(Account<'a>),
+    Amount(Amount<'a>),
+    Currency(String),
+    Date(Date),
+    String(String),
+    // FIXME: Do not yet handle:
+    // 1) Empty metadata
+    // 2) Tags
+    // 3) Numbers without currencies
+}
+
+fn metadata_key(input: &str) -> IResult<&str, &str> {
+    recognize(pair(
+        satisfy(|c: char| c.is_ascii_lowercase() && c.is_ascii_alphabetic()),
+        many0(alt((
+            satisfy(|c: char| c.is_ascii_alphabetic()),
+            satisfy(char::is_numeric),
+            one_of("-_"),
+        ))),
+    ))(input)
+}
+
+fn metadata_value(input: &str) -> IResult<&str, Value<'_>> {
+    alt((
+        map(account, Value::Account),
+        map(amount, Value::Amount),
+        map(currency, |c| Value::Currency(c.to_owned())),
+        map(date, Value::Date),
+        map(string, Value::String),
+    ))(input)
+}
+
+fn metadata_line(input: &str) -> IResult<&str, (&str, Value<'_>)> {
+    separated_pair(
+        metadata_key,
+        tuple((space0, tag(":"), space0)),
+        metadata_value,
+    )(input)
+}
+
+pub fn metadata(input: &str) -> IResult<&str, Metadata<'_>> {
```

Comment:
Make sure, that won't accidentally be accessible from outside the crate.
```suggestion
pub(crate) fn metadata(input: &str) -> IResult<&str, Metadata<'_>> {
```